### ISSUE TITLE
Support locking window

### ISSUE THREAD

**Issue by Keith-CY** [Magickbase/neuron-public-issues#318]
This feature request comes from the community.

---

The case is that Neuron may be running on a communal computer of a group or friends while the owner wants to keep Neuron private. It's not feasible now because Neuron doesn't require authorization to access to balance/history.

So the feature required is that the window could be locked actively by the owner, and only **unlock** is allowed when the window is locked.

---

This feature is just an enhancement, but necessary or of high-priority.

**Comment by Danie0918** [comment]
【Optimization of hot/cold combinations for wallets】

- Requirement Description: The case is that Neuron may be running on a communal computer of a group of friends while the owner wants to keep Neuron private.

![support_locking_window](https://github.com/Magickbase/neuron-public-issues/assets/122437870/d27c1aeb-be9a-440a-9b0f-63ecc2416dbb)

- Requirement Contents
1. Add the Lock Window option to the Window menu. (Set Comand/Ctrl + L as the shortcut key)
2. Only window adjustment and closing are supported when the Neuron is locked.
3. In the locked state, click the Unlock button and enter the password to unlock it.

**Comment by Keith-CY** [comment]
Maybe we can use the native dialog for unlocking
![image](https://github.com/Magickbase/neuron-public-issues/assets/7271329/e9a86a69-31e2-40bd-abe9-8a2e09c6942d)

BTW, will there be an independent password for unlocking? I think the security level should be different.

**Comment by Danie0918** [comment]
> BTW, will there be an independent password for unlocking? I think the security level should be different.

Currently designed to be a universal password, an independent password could perhaps be introduced when there are more requirements. This is because an independent password means additional memorization and management.

**Comment by yanguoyu** [comment]
If someone restarts the Neuron wallet, does it should still locked if Neuron is locked before quitting?

**Comment by Danie0918** [comment]
> If someone restarts the Neuron wallet, does it should still locked if Neuron is locked before quitting?

Remains locked until it is unlocked.

**Comment by Kirl70** [comment]
Design Draft (0-1--0-5):
https://www.figma.com/file/6XNoimRDbFTTNm016rbIdU/Magickbase?type=design&node-id=32213%3A56909&mode=design&t=KnCUMYYLePbTRJQt-1
<img width="1446" alt="image" src="https://github.com/Magickbase/neuron-public-issues/assets/106060898/a1c6f51c-b290-4b7b-bcf8-354923ded455">

**Comment by Keith-CY** [comment]
I've found a detail that may be confusing

<img width="1201" alt="image" src="https://github.com/Magickbase/neuron-public-issues/assets/7271329/8c83cda3-0d82-450c-9b9e-e262c2511041">

The menu item is `lock window` but the information is `wallet is locked`.

I think `lock window` and `window is locked` are clearer because `lock wallet` is a bit scary, and I would worry if the synchronization stops.

---

And as mentioned at https://github.com/Magickbase/neuron-public-issues/issues/318#issuecomment-1873679138, I'd suggest using an independent password for window unlocking because it's of a lower level of security. I would keep my wallet password used as little as possible.

**Comment by Danie0918** [comment]
> > Currently designed to be a universal password, an independent password could perhaps be introduced when there are more requirements. This is because an independent password means additional memorization and management.
> 
> The password to unlock window could be relatively simple, such as a PIN code, that doesn't require much memorization. Besides, the wallet can be backed up technically, so it won't block users if the PIN code is lost.

Does this mean that Neuron needs to add a PIN setup process or can Neuron use the system PIN?

**Comment by Keith-CY** [comment]
> > > Currently designed to be a universal password, an independent password could perhaps be introduced when there are more requirements. This is because an independent password means additional memorization and management.
> > 
> > 
> > The password to unlock window could be relatively simple, such as a PIN code, that doesn't require much memorization. Besides, the wallet can be backed up technically, so it won't block users if the PIN code is lost.
> 
> Does this mean that Neuron needs to add a PIN setup process or can Neuron use the system PIN?

It might be impossible to get system pin code. PIN code is just an example, the point is that users may use a simple code to protect the window, while use a complex password to protect the wallet.

Besides, when it's the hardware wallet, user has to expose the password on the computer if the wallet password is used in this case.

**Comment by Kirl70** [comment]
To make the page more dynamic, I suggest placing the password input box on the outer layer. 
Demo:
Unlock

https://github.com/Magickbase/neuron-public-issues/assets/106060898/8f8a082e-4ab5-4dde-bb25-ea3090be234b


Wrong Password


https://github.com/Magickbase/neuron-public-issues/assets/106060898/7d9062d1-efea-4321-94af-6c27743a58d1

**Comment by Danie0918** [comment]
![drawing_unlock](https://github.com/Magickbase/neuron-public-issues/assets/122437870/deaad379-ee8e-4c08-b3da-ee13b262bee9)


To differentiate passwords and reduce the cost of memorization for the user, consider using a pattern password.

- Requirement Contents
1. Add the Lock Window option to the Window menu. (Set Comand/Ctrl + L as the shortcut key)
    - When the user uses the lock window for the first time, open the pattern password setting and lock the window when finished.
2. Add the setting pattern password in settings.
    - First time setting, display 'Set unlock pattern'.
    - Set it again to show 'Modify unlock pattern' （Requires verification of the set pattern password）
3. Only window adjustment and closing are supported when the Neuron is locked.
4. In the locked state, click the Unlock button and enter the password to unlock it.
    - Draw the pattern password successfully to unlock the window.
    - The retry time is increased to 5 minutes, 30 minutes, and 24 hours in order if the pattern code fails more than a multiple of 3 times during the same locking window period. (Maximum is 24 hours, failing again more than 3 times still locks for 24 hours)

**Comment by yanguoyu** [comment]
why not add a four-number password to lock and unlock? Is it too complex to use `pattern` on the computer?

**Comment by Keith-CY** [comment]
> why not add a four-number password to lock and unlock? Is it too complex to use `pattern` on the computer?

I agree that `gesture pattern` is too complex on a computer unless the screen is touchable.

**Comment by Danie0918** [comment]
> > why not add a four-number password to lock and unlock? Is it too complex to use `pattern` on the computer?
> 
> I agree that `gesture pattern` is too complex on a computer unless the screen is touchable.

This only requires a long press on the mouse, though it may be a little harder with a trackpad.

For four-number password, it may create memory confusion with the original password, I'll think about it.

**Comment by Danie0918** [comment]
![pin](https://github.com/Magickbase/neuron-public-issues/assets/122437870/1645ac89-0903-4596-b00b-9c7371bd04a1)

- Requirement Contents
1. Add the Lock Window option to the Window menu. (Set Comand/Ctrl + L as the shortcut key)
    - When the user uses the lock window for the first time, open the unlock code setting and lock the window when finished.
2. Add the setting unlock code in settings.
    - First time setting, display 'Set unlock code'.
    - Set it again to show 'Modify unlock code' （Requires verification of the set unlock code）
3. Only window adjustment and closing are supported when the Neuron is locked.
4. In the locked state, click the Unlock button and enter the unlock code to unlock it.
    - Enter the unlock code successfully to unlock the window.
    - The retry time is increased to 5 minutes, 30 minutes, and 24 hours in order if the unlock code fails more than a multiple of 3 times during the same locking window period. (Maximum is 24 hours, failing again more than 3 times still locks for 24 hours)

**Comment by Kirl70** [comment]
Design Draft:
https://www.figma.com/file/6XNoimRDbFTTNm016rbIdU/Magickbase?type=design&node-id=32573%3A34064&mode=design&t=ynLL7fvmguFtOzLw-1
<img width="761" alt="image" src="https://github.com/Magickbase/neuron-public-issues/assets/106060898/99bf5017-1a3d-412b-8794-8c0acd0839e0">

**Comment by yanguoyu** [comment]
https://github.com/nervosnetwork/neuron/pull/3064
